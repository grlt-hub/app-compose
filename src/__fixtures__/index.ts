import { nanoid } from 'nanoid';

const genContainerId = () => nanoid(7) as string & { __minLength: 1 };

export { genContainerId };
