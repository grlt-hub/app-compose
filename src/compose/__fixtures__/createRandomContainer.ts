import { createContainer, type AnyContainer } from '@createContainer';
import { randomUUID } from 'node:crypto';

type Overrides = Partial<Pick<AnyContainer, 'domain' | 'dependsOn' | 'optionalDependsOn' | 'id' | 'enable' | 'start'>>;

export const createRandomContainer = (overrides: Overrides = {}): AnyContainer =>
  createContainer({
    // @ts-expect-error
    id: randomUUID(),
    domain: randomUUID(),
    start: () => ({ api: null }),
    enable: () => true,
    ...overrides,
  });
