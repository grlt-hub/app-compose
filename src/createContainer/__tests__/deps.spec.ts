import { genContainerId } from '../../__fixtures__';
import { createContainer } from '../index';

const start = () => ({ api: {} });

const __ = {
  a: createContainer({
    id: 'a',
    start: () => ({ api: { t: () => true } }),
  }),
  b: createContainer({
    id: 'b',
    start: () => ({ api: { f: () => false } }),
  }),
  c: createContainer({
    id: 'd',
    start: () => ({ api: { nil: null } }),
  }),
};

describe('container deps are uniq', () => {
  test('happy', () => {
    expect(() =>
      createContainer({ id: genContainerId(), start, dependsOn: [__.a], optionalDependsOn: [__.b] }),
    ).not.toThrowError();
    expect(() => createContainer({ id: genContainerId(), start, dependsOn: [__.a] })).not.toThrowError();
    expect(() => createContainer({ id: genContainerId(), start, optionalDependsOn: [__.b] })).not.toThrowError();
    expect(() => createContainer({ id: genContainerId(), start })).not.toThrowError();
  });

  test('unhappy', () => {
    const id = 'pu-pu-pu';

    {
      expect(() =>
        createContainer({ id, start, dependsOn: [__.a], optionalDependsOn: [__.a] }),
      ).toThrowErrorMatchingSnapshot();
    }

    {
      expect(() =>
        createContainer({ id, start, dependsOn: [__.a, __.b, __.c], optionalDependsOn: [__.b] }),
      ).toThrowErrorMatchingSnapshot();
    }

    {
      expect(() =>
        createContainer({ id, start, dependsOn: [__.b], optionalDependsOn: [__.a, __.b, __.c] }),
      ).toThrowErrorMatchingSnapshot();
    }
  });
});
