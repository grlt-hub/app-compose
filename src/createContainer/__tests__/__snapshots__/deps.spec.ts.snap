// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`container deps are uniq > unhappy 1`] = `
[Error: 
      Dependency conflict detected in container 'pu-pu-pu':
      The following dependencies are listed as both required and optional: [a].

      Each dependency should be listed only once, as either required or optional.]
`;

exports[`container deps are uniq > unhappy 2`] = `
[Error: 
      Dependency conflict detected in container 'pu-pu-pu':
      The following dependencies are listed as both required and optional: [b].

      Each dependency should be listed only once, as either required or optional.]
`;

exports[`container deps are uniq > unhappy 3`] = `
[Error: 
      Dependency conflict detected in container 'pu-pu-pu':
      The following dependencies are listed as both required and optional: [b].

      Each dependency should be listed only once, as either required or optional.]
`;
